### ISSUE TITLE
When opening a document many times, some users stay at "0% connecting"

### ISSUE THREAD

**Issue by Frando** [nextcloud/richdocuments#2125]
**Describe the bug**

I am using Nextcloud Office/Richdocuments with `collabora/code:21.11.3.4.1`, both installed via Docker. The setup works well, until a few users start to edit the same document. At around 5 or 6 users, people start to see a message "Nextcloud office / 0% connecting", and no further changes. The first users can continue to edit the document.

![nextcloud](https://user-images.githubusercontent.com/43627/161731826-e9e54571-bad8-4452-a1df-861e800d4bdf.png)

Interestingly, when users connect or disconnect, the user bubbles still appear and disappear. Also, the websocket connection to collabora is established, and I see the same pattern of messages as in the working editing session.
![nextcloud2](https://user-images.githubusercontent.com/43627/161732703-3ffc7fe7-963e-4f0a-b8b0-a0a92d1d141b.png)

So it seems the connection to Collabora is established and seems to be working. Also, in the Collabora admin panel, the number of connections includes the sessions that stay at "0% connecting". I also couldn't find any errors in the Collabora logs.

On the tabs that stayed at "0% connecting", the browser console log contained one error that was not present in the logs of the working tabs.
```
Uncaught TypeError: app.sectionContainer.getSectionWithName(...) is null
    _hideWizard https://collabora.rdlnet.de/browser/2ea9477/bundle.js:1
    onAdd https://collabora.rdlnet.de/browser/2ea9477/bundle.js:1
    addControl https://collabora.rdlnet.de/browser/2ea9477/bundle.js:1
    initializeBasicUI https://collabora.rdlnet.de/browser/2ea9477/bundle.js:1
    <anonymous> https://collabora.rdlnet.de/browser/2ea9477/bundle.js:1
    <anonymous> https://collabora.rdlnet.de/browser/2ea9477/bundle.js:1
    <anonymous> https://collabora.rdlnet.de/browser/2ea9477/bundle.js:1
bundle.js:1:2270506
```
See below for full logs.

**To Reproduce**
1. Install Nextcloud and Collabora via docker-compose and configure according to instructions
2. Start opening the same document many times (~6+)
3. Experience that the first documents open as intendended, and from the 6th or 7th on they stay at "0% connecting"

**Expected behavior**
All documents open, or the Collabora logs would reveal an error, or the system would run out of RAM.

**Screenshots**
See above

**Client details:**
 - OS: Linux
 - Browser Firefox
 - Version 98
 - Device: Desktop

## Server details

**Operating system**: VM with Debian Linux, Nextcloud hosted in Docker Compose. The VM as 12GB of memory, and the collabora container uses around 800MB-1GB usually.

**Web server:** Caddy, forwards to the php-fpm image and reverse-proxy for collabora, everything with HTTPS over letsencrypt (all working)

**Database:** mariadb, latest Docker image for tag `mariadb`

**PHP version:** `8.0.15` (via `nextcloud:fpm` image)

**Nextcloud version:**
Docker image `nextcloud:23.0.3-fpm`

**Version of the richdocuments app**
`5.0.3.`

**Version of Collabora Online**
Docker image `collabora/code:21.11.3.4.1`

<details>
<summary>Logs</summary>

#### Nextcloud log (data/nextcloud.log)

Here's the log from docker-compose for both nextcloud and collabora.
[log2.txt](https://github.com/nextcloud/richdocuments/files/8417229/log2.txt)

#### Browser log
```
Here's the console log for a tab that stayed at "0% connecting"

Opening viewer for file  /Documents/Welcome to Nextcloud Hub.docx Viewer.vue:259
Fetching additional files... Viewer.vue:297
OC.getCapabilities is deprecated and will be removed in Nextcloud 21. See @nextcloud/capabilities capabilities.js:35:9
Content Security Policy: Ignoring “'unsafe-inline'” within script-src or style-src: nonce-source or hash-source specified
Content Security Policy: Ignoring ‘x-frame-options’ because of ‘frame-ancestors’ directive.
Content Security Policy: Ignoring “'unsafe-inline'” within script-src or style-src: nonce-source or hash-source specified
No OC found index.js:46:12
Proxying an event bus of version 2.1.1 with 1.3.0 index.es.js:2337:14
JQMIGRATE: Migrate is installed, version 3.3.2 jquery-migrate.min.js:2:708
Proxying an event bus of version 2.1.1 with 1.3.0 index.es.js:2337:14
session heartbeat polling started session-heartbeat.js:101:9
PostMessageService.sendPostMessage parent loading postMessage.tsx:57:10
Waiting for page to render… document.js:219:12
Content Security Policy: The page’s settings blocked the loading of a resource at inline (“script-src”). moz-extension:221:46
PostMessageService.sendPostMessage parent loading postMessage.tsx:57:10
[getWopiUrl] https://cloud.rdlnet.de/index.php/apps/richdocuments/wopi/files/1627_ocsjoizdgw0c url.js:39:9
Content Security Policy: Ignoring “'unsafe-inline'” within script-src or style-src: nonce-source or hash-source specified
Content Security Policy: Ignoring “'unsafe-inline'” within script-src or style-src: nonce-source or hash-source specified 5
Content Security Policy: Ignoring “'unsafe-inline'” within script-src or style-src: nonce-source or hash-source specified
This page uses the non standard property “zoom”. Consider using calc() in the relevant property values, or using “transform” along with “transform-origin: 0 0”. cool.html:107:41
Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user’s experience. For more help http://xhr.spec.whatwg.org/ bundle.js:1:766250
PostMessageService.sendPostMessage loolframe {"MessageId":"Host_PostmessageReady","SendTime":1649154664351,"Values":{}} postMessage.tsx:57:10
[document] editorInitListener: Received post message  
Object { msgId: "App_LoadingStatus", args: {…}, deprecated: false }
document.js:259:13
PostMessageService.sendPostMessage parent {"MessageId":"App_LoadingStatus","SendTime":1649154664378,"Values":{"Status":"Frame_Ready","Features":{"VersionStates":true}}} postMessage.tsx:57:10
callMobileMessage documentLoaded undefined mobile.js:35:9
[document] Received post message  
Object { msgId: "App_LoadingStatus", args: {…}, deprecated: false }
document.js:339:13
PostMessageService.sendPostMessage parent {"MessageId":"App_LoadingStatus","SendTime":1649154664378,"Values":{"Status":"Frame_Ready","Features":{"VersionStates":true}}} postMessage.tsx:57:10
[document] Unhandled post message 
Object { msgId: "App_LoadingStatus", args: {…}, deprecated: false }
document.js:439:14
[document] editorInitListener: Received post message  
Object { msgId: "Action_Load_Resp", args: {…}, deprecated: false }
document.js:259:13
PostMessageService.sendPostMessage parent {"MessageId":"Action_Load_Resp","SendTime":1649154664379,"Values":{"success":true,"result":"","errorMsg":""}} postMessage.tsx:57:10
[document] Received post message  
Object { msgId: "Action_Load_Resp", args: {…}, deprecated: false }
document.js:339:13
PostMessageService.sendPostMessage parent {"MessageId":"Action_Load_Resp","SendTime":1649154664379,"Values":{"success":true,"result":"","errorMsg":""}} postMessage.tsx:57:10
[document] Unhandled post message 
Object { msgId: "Action_Load_Resp", args: {…}, deprecated: false }
document.js:439:14
PostMessageService.handlePostMessage Ignoring deprecated post message View_Added 26 postMessage.tsx:111:12
[document] editorInitListener: Received post message  
Object { msgId: "Views_List", args: (13) […], deprecated: false }
document.js:259:13
[document] Received post message  
Object { msgId: "Views_List", args: (13) […], deprecated: false }
document.js:339:13
PostMessageService.sendPostMessage parent {"MessageId":"Views_List","SendTime":1649154664524,"Values":[{"ViewId":"18","UserName":"Franz","UserId":"keycloak.rdlnet.de-df214dea-eab9-4543-8a9a-abaed8df55c9","UserExtraInfo":{"avatar":"https://cloud.rdlnet.de/avatar/keycloak.rdlnet.de-df214dea-eab9-4543-8a9a-abaed8df55c9/32","is_admin":true},"Color":411298,"ReadOnly":"0","IsCurrentView":false},{"ViewId":"20","UserName":"Franz","UserId":"keycloak.rdlnet.de-df214dea-eab9-4543-8a9a-abaed8df55c9","UserExtraInfo":{"avatar":"https://cloud.rdlnet.de/avatar/keycloak.rdlnet.de-df214dea-eab9-4543-8a9a-abaed8df55c9/32","is_admin":true},"Color":411298,"ReadOnly":"0","IsCurrentView":false},{"ViewId":"22","UserName":"Franz","UserId":"keycloak.rdlnet.de-df214dea-eab9-4543-8a9a-abaed8df55c9","UserExtraInfo":{"avatar":"https://cloud.rdlnet.de/avatar/keycloak.rdlnet.de-df214dea-eab9-4543-8a9a-abaed8df55c9/32","is_admin":true},"Color":411298,"ReadOnly":"0","IsCurrentView":false},{"ViewId":"24","UserName":"Franz","UserId":"keycloak.rdlnet.de-df214dea-eab9-4543-8a9a-abaed8df55c9","UserExtraInfo":{"avatar":"https://cloud.rdlnet.de/avatar/keycloak.rdlnet.de-df214dea-eab9-4543-8a9a-abaed8df55c9/32","is_admin":true},"Color":411298,"ReadOnly":"0","IsCurrentView":false},{"ViewId":"26","UserName":"Franz","UserId":"keycloak.rdlnet.de-df214dea-eab9-4543-8a9a-abaed8df55c9","UserExtraInfo":{"avatar":"https://cloud.rdlnet.de/avatar/keycloak.rdlnet.de-df214dea-eab9-4543-8a9a-abaed8df55c9/32","is_admin":true},"Color":411298,"ReadOnly":"0","IsCurrentView":false},{"ViewId":"28","UserName":"Franz","UserId":"keycloak.rdlnet.de-df214dea-eab9-4543-8a9a-abaed8df55c9","UserExtraInfo":{"avatar":"https://cloud.rdlnet.de/avatar/keycloak.rdlnet.de-df214dea-eab9-4543-8a9a-abaed8df55c9/32","is_admin":true},"Color":411298,"ReadOnly":"0","IsCurrentView":false},{"ViewId":"29","UserName":"Franz","UserId":"keycloak.rdlnet.de-df214dea-eab9-4543-8a9a-abaed8df55c9","UserExtraInfo":{"avatar":"https://cloud.rdlnet.de/avatar/keycloak.rdlnet.de-df214dea-eab9-4543-8a9a-abaed8df55c9/32","is_admin":true},"Color":411298,"ReadOnly":"0","IsCurrentView":false},{"ViewId":"31","UserName":"Franz","UserId":"keycloak.rdlnet.de-df214dea-eab9-4543-8a9a-abaed8df55c9","UserExtraInfo":{"avatar":"https://cloud.rdlnet.de/avatar/keycloak.rdlnet.de-df214dea-eab9-4543-8a9a-abaed8df55c9/32","is_admin":true},"Color":411298,"ReadOnly":"0","IsCurrentView":false},{"ViewId":"32","UserName":"Franz","UserId":"keycloak.rdlnet.de-df214dea-eab9-4543-8a9a-abaed8df55c9","UserExtraInfo":{"avatar":"https://cloud.rdlnet.de/avatar/keycloak.rdlnet.de-df214dea-eab9-4543-8a9a-abaed8df55c9/32","is_admin":true},"Color":411298,"ReadOnly":"0","IsCurrentView":false},{"ViewId":"33","UserName":"Franz","UserId":"keycloak.rdlnet.de-df214dea-eab9-4543-8a9a-abaed8df55c9","UserExtraInfo":{"avatar":"https://cloud.rdlnet.de/avatar/keycloak.rdlnet.de-df214dea-eab9-4543-8a9a-abaed8df55c9/32","is_admin":true},"Color":411298,"ReadOnly":"0","IsCurrentView":false},{"ViewId":"34","UserName":"Franz","UserId":"keycloak.rdlnet.de-df214dea-eab9-4543-8a9a-abaed8df55c9","UserExtraInfo":{"avatar":"https://cloud.rdlnet.de/avatar/keycloak.rdlnet.de-df214dea-eab9-4543-8a9a-abaed8df55c9/32","is_admin":true},"Color":411298,"ReadOnly":"0","IsCurrentView":false},{"ViewId":"36","UserName":"Franz","UserId":"keycloak.rdlnet.de-df214dea-eab9-4543-8a9a-abaed8df55c9","UserExtraInfo":{"avatar":"https://cloud.rdlnet.de/avatar/keycloak.rdlnet.de-df214dea-eab9-4543-8a9a-abaed8df55c9/32","is_admin":true},"Color":411298,"ReadOnly":"0","IsCurrentView":false},{"ViewId":"37","UserName":"Franz","UserId":"keycloak.rdlnet.de-df214dea-eab9-4543-8a9a-abaed8df55c9","UserExtraInfo":{"avatar":"https://cloud.rdlnet.de/avatar/keycloak.rdlnet.de-df214dea-eab9-4543-8a9a-abaed8df55c9/32","is_admin":true},"Color":411298,"ReadOnly":"0","IsCurrentView":false}]} postMessage.tsx:57:10
[document] Unhandled post message 
Object { msgId: "Views_List", args: (13) […], deprecated: false }
document.js:439:14
[document] editorInitListener: Received post message  
Object { msgId: "Doc_ModifiedStatus", args: {…}, deprecated: false }
document.js:259:13
[document] Received post message  
Object { msgId: "Doc_ModifiedStatus", args: {…}, deprecated: false }
document.js:339:13
PostMessageService.sendPostMessage parent {"MessageId":"Doc_ModifiedStatus","SendTime":1649154664544,"Values":{"Modified":false}} postMessage.tsx:57:10
[document] Unhandled post message 
Object { msgId: "Doc_ModifiedStatus", args: {…}, deprecated: false }
document.js:439:14
[viewer] Received post message 
Object { msgId: "App_LoadingStatus", args: {…}, deprecated: false }
Office.vue:221
[viewer] Received post message 
Object { msgId: "App_LoadingStatus", args: {…}, deprecated: false }
Office.vue:221
[viewer] Received post message 
Object { msgId: "Action_Load_Resp", args: {…}, deprecated: false }
Office.vue:221
Uncaught TypeError: app.sectionContainer.getSectionWithName(...) is null
    _hideWizard https://collabora.rdlnet.de/browser/2ea9477/bundle.js:1
    onAdd https://collabora.rdlnet.de/browser/2ea9477/bundle.js:1
    addControl https://collabora.rdlnet.de/browser/2ea9477/bundle.js:1
    initializeBasicUI https://collabora.rdlnet.de/browser/2ea9477/bundle.js:1
    <anonymous> https://collabora.rdlnet.de/browser/2ea9477/bundle.js:1
    <anonymous> https://collabora.rdlnet.de/browser/2ea9477/bundle.js:1
    <anonymous> https://collabora.rdlnet.de/browser/2ea9477/bundle.js:1
bundle.js:1:2270506
[viewer] Received post message 
Object { msgId: "Action_Load_Resp", args: {…}, deprecated: false }
Office.vue:221
[viewer] Received post message 
Object { msgId: "Views_List", args: (13) […], deprecated: false }
Office.vue:221
[viewer] Received post message 
Object { msgId: "Doc_ModifiedStatus", args: {…}, deprecated: false }
Office.vue:221


```

</details>

**Comment by danir-de** [comment]
Hey, this is a known bug, that was first discussed here: https://github.com/nextcloud/richdocuments/issues/2017#issuecomment-1054095255

I can confirm, that this only happens in Firefox, Chromium based browsers seem perfectly fine.

**Comment by Frando** [comment]
Is there any path to resolution here? This is actively blocking us from using Collabora in Nextcloud and we're considering having to move away from Nextcloud/Collabora in practice, due to this issue. I'm asking because the linked issue is closed, and supporting several users editing the same document for me seemed to be one of the main premises of the Collabora integration in Nextcloud.

**Comment by danir-de** [comment]
We're having a support ticket with collabora and they say that they could reproduce it on our setup a few days ago.

**Comment by alangecker** [comment]
**TLDR:** a hotfix is at the end

after a deep dive into the rabbit hole I think I got the bug!

it was now an effort of two days and opening the same `.odt` document in nextcloud like a million times with multiple "wtf?" moments, why I want to share some bits of the story :D 

## Debugging Collabora

After a while following some unrelated log messages, I've compared the websocket traffic and found out, that the messages sent from the server are the same, but there are no messages from the client send to the server to further initialize the application.

usually they seem to be sent by `Map.sendInitUNOCommands()` within this stack
![image](https://user-images.githubusercontent.com/10246027/168418651-d36070f7-5aa9-4ef0-87e4-9250351aca04.png)

after a while of digging around in the functions in the stack and understanding where it decides to not execute the branch which would lead to `sendInitUNOCommands()`, I got to:

`L.CanvasTileLayer.onAdd(map)`
```js
	onAdd: function (map) {
		[...]
		this._initContainer();
		[...]
		this._map.sendInitUNOCommands();
		[...]
	},
```

in the cases where it doesn't work, something strange happens:

### Browser bug?! `ctx.font = `
within `_initContainer()` the method `CanvasSectionContainer.drawSections` calls something like
```js
ctx.font = "17px Verdana"
```
BUT this command never finishes. no endless loop. no high cpu usage. no error thrown. the event loop continues elsewhere as usual, but all further code in this stack doesn't continue. As far as I understand javascript, this is/should be not possible! it must be a bug in the js/dom engine!? but same behavior in firefox and chromium - different source -> wired 🤔 

I've verified this with different strings / font families & sizes, but always the same behavior
![image](https://user-images.githubusercontent.com/10246027/168418471-10029bcc-4dd5-4c7e-9403-a4e5ea023cb3.png)
![image](https://user-images.githubusercontent.com/10246027/168418392-9359b635-8357-4cb0-ab09-e1c097a520c4.png)

this must be the bug! I've spent a while in the [firefox C++ source](https://github.com/bolucat/Firefox/blob/f6f915cb6b56045d2b44cacccbb042579ec2a8f9/dom/canvas/CanvasRenderingContext2D.cpp#L3395=), trying to understand what `ctx.font =` does and why it doesn't return - but yeah... thats a completely different path in the rabbit hole, which I luckily gave up at some point.

Not continuing leads collabora to be in a half initialized state. I've disabled the line, but then the same behavior within similar canvas context calls. after hours of modifications I couldn't manage it to further initialize - too many interactions with the context.

but why does it only happen when the document is open multiple times? what is different?

### maybe it's not our bug...
I started to believe, that the canvas bug is not the origin of this issue, only a side effect of something else being wired. playing around for another (longer) while of browser debugging, modifying code and loaads of document re-openings, I noticed, that `CanvasSectionContainer.onResize(newWidth: number, newHeight: number)` is in the working case always called with numbers >0, but in the error case always called firstly with (0,0) and after some seconds with "normal" numbers >0.

-> the final hint to the right path!

### canvas size == 0
seems like somehow the canvas has an height and width of 0, which makes `ctx.font =` unhappy?

but why is it initially zero when the document is opened multiple times? how is the size defined? took me again a while of staring at JS code to realize again that there is also CSS that does the things that CSS does - crazy!

ultimately the canvas size is zero because the whole page/iframe is zero, it is set to flex and there is (in that moment) no height/width yet specified.

but how does the app define the size of the iframe?

## Debugging Nextcloud Office

in the `Office.vue` we find out, that the iframe is loaded hidden (`display: none`) and only shown (and apparently given a size) when the loading state is at least `FRAME_READY`

```js
showIframe() {
	return this.loading >= LOADING_STATE.FRAME_READY
},
```

and that state is set, when we get a `App_LoadingStatus` message with `Status === 'Frame_Ready'` from collabora - okedoki!

## Back in Collabora
good, that the collabora code feels already a bit familiar :D

`Frame_Ready` is sent via postMessage by the collabora client when it receives a `wopi:` from the server, which happens in all cases.... but...  at different times!!

## seems like the whole point is
when other clients are connected, the messages from the server take a bit longer, which causes the Collabora client to initialize even before the iframe gets a size and thus, due to the strange browser bug, the initialization fails..

## Hotfix

in the bundle `/apps/richdocuments/js/richdocuments-viewer.js` there is 

```js 
showIframe:function(){return this.loading>=ue}
```

I simple changed it to always show the iframe:
```js
showIframe:function(){return true}
```

not beautiful, but it reliably fixes the issue for me after a long journey :)

**Comment by danir-de** [comment]
> not beautiful, but it reliably fixes the issue for me after a long journey :)

Wow great job!
We'll try this out in production next week! (Can't really make it worse atm 😃)

**Comment by XueSheng-GIT** [comment]
> I simple changed it to always show the iframe:
> 
> ```js
> showIframe:function(){return true}
> ```

@alangecker thanks a lot for your efforts!
I changed the mentioned line in my nextcloud setup, but still get the same error. Anything else which needs to be done to make it work?

**Comment by juliushaertl** [comment]
Wow, that is really impressive debugging, thanks for sharing all those insights. I'll discuss this with Collabora early next week to see how we can make this work or if we need to avoid the loading while the iframe is hidden on the Nextcloud side.

**Comment by XueSheng-GIT** [comment]
After some testing it seems that this hotfix also resolves the issue that Nextcloud Office only opens with an empty page from time to time.

References: https://github.com/nextcloud/richdocuments/issues/2017 (closed, but actually not solved!) and https://github.com/nextcloud/richdocuments/issues/2094

**Comment by juliushaertl** [comment]
I've pushed a potential fix in #2212, if anyone is up for help testing this, this would be highly appreciated. It can be hot patched in a similar way as the hot fix mentioned earlier.

**Comment by XueSheng-GIT** [comment]
@juliushaertl tried your hotpatch, but document can't be opened afterwards (spinner is shown forever) and I get the following error in the console of firefox:
> ReferenceError: _vm is not defined

**Comment by juliushaertl** [comment]
Oh my bad then I mixed up the patch with the minified version. Depending on the version you may need to use `e.showIframe` as it is used in the replaced code part.

**Comment by juliushaertl** [comment]
Updated the hot patch description accordingly.

**Comment by XueSheng-GIT** [comment]
Thanks the explanation regarding the variable! Tested on two instances with NC23 and NC24. Opening docs within firefox as single user or with multiple users works without any issues so far! Thanks a lot for fixing that issue!
